Extending the C library and the python interface

evoyiatzis opened the pull request:
**Summary**

Extending the C library and the python interface by adding three new methods for gathering angle, dihedral and improper information

**Related Issue(s)**

No related issue

**Author(s)**

E. Voyiatzis - email: evoyiatzis@gmail.com

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

it does not break backward compatibility 

**Implementation Notes**

Correctness was verified by extracting angle, dihedral and improper info from the united_atom_polystyrene.txt file using the following commands in the python command prompt:

from lammps import lammps
lmp=lammps()
lmp.command("atom_style full")
lmp.command("read_data united_atom_polystyrene.data nocoeff")
nang, topology_1 = lmp.gather_angles()
ndih, topology_2 = lmp.gather_dihedrals()
nimp, topology_3 = lmp.gather_impropers()

and checking the obtained values against the ones in the data file.

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [X] The feature or features in this pull request is complete
- [X] Licensing information is complete
- [X] Corresponding author information is complete
- [X] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [X] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

[united_atom_polystyrene.txt.gz](https://github.com/lammps/lammps/files/10502355/united_atom_polystyrene.txt.gz)

evoyiatzis committed:
Update library.h

evoyiatzis committed:
Update library.cpp

evoyiatzis committed:
Update numpy_wrapper.py

evoyiatzis committed:
Update core.py

akohlmey commented:
@evoyiatzis thanks for your submission. We should try to extend this change to also be included in the LAMMPS Fortran module and the LAMMPS Swig interface description, so that other supported wrappers of the library interface have the same functionality.

I can help with that. Perhaps @hammondkd may have some suggestions for how to best implement the Fortran part. Now that he managed to have "feature parity" for the Fortran module, we should make an effort to keep that in sync.
The support for Swig is rather simple, it mostly needs updates the interface description in tools/swig/lammps.i

akohlmey commented on python/lammps/core.py:
```diff
@@ -1299,6 +1308,75 @@ def gather_bonds(self):
+    flat list of ctypes integer values with the angle type, angle atom1,
+    angle atom2, angle atom3 for each angle.
+
+    .. versionadded:: 28Jan2023
```
The version number for versionadded/versionchanged should be "TBD". This applies to all similar tags in the remaining changes. The TBD will be replaced with the actual version in the last pull request before the release that also updates src/version.h

akohlmey commented on src/library.cpp:
```diff
@@ -3159,6 +3159,114 @@ void lammps_gather_bonds(void *handle, void *data)
 }
 
+// Gather function for angles
+void lammps_gather_angles(void *handle, void *data)
```
There need to be doxygen comments added here to document the function. Same for the others below.

evoyiatzis replied:
I think I have added now the required doxygen comments in the library.cpp

akohlmey reviewed (changes requested):
Here is a TODO list of all needed changes to make the additions here consistent with the rest of the library functions. Adding to the library interface is becoming a bit of a burden since there are now so many uses of it :cry:. After it took a lot of effort and energy to bring everything to the same state, we now want to keep it like that.

It should not be too difficult to do the updates in this case, since everything can follow the example of the `lammps_gather_bonds()` function.

Please let us know which tasks you can/want to do yourself and where you need help.

- [x] add Fortran wrappers to `fortran/lammps.f90`
- [x] add doxygen style comments to `src/library.cpp` to document the new C library functions added
- [x] add the new functions and prototypes to `examples/COUPLE/plugin/liblammpsplugin.h` and `examples/COUPLE/plugin/liblammpsplugin.c`
- [x] add the new function prototypes to `tools/swig/lammps.i`
- [x] add entries that reference these comments/functions to `doc/src/Library_scatter.rst`
- [x] document the Fortran module calls in `doc/src/Fortran.rst`
- [x] add unit tests for the C library calls to `unittest/c-library/test_library_scatter_gather.cpp`
- [x] add unit tests for the python functions to `unittest/python/python-scatter-gather.py`
- [x] add unit tests for the numpy wrapper functions to `unittest/python/python-numpy.py`
- [x] add unit tests for the new Fortran module calls to `unittest/fortran/test_fortran_gather_scatter.f90` and `unittest/fortran/wrap_gather_scatter.cpp`

evoyiatzis committed:
Update lammps.i

evoyiatzis committed:
Update liblammpsplugin.h

evoyiatzis committed:
Update liblammpsplugin.c

evoyiatzis committed:
Update Library_scatter.rst

evoyiatzis committed:
add doxygen style comments to document the new C library functions added

evoyiatzis committed:
add Fortran wrappers to fortran/lammps.f90

evoyiatzis committed:
document the Fortran module calls

evoyiatzis committed:
Auxiliary test methods for angles, dihedrals and impropers

evoyiatzis committed:
Auxiliary test methods for angles, dihedrals and impropers in python-numpy

evoyiatzis committed:
Replacing arbitrary LAMMPS version with TBD in core.py

evoyiatzis committed:
Replacing arbitrary LAMMPS version with TBD in numpy_wrapper.py

evoyiatzis committed:
Include tests for gather_angles, gather_dihedrals & gather_impropers in python-numpy.py

evoyiatzis committed:
Include tests for gather_angles, gather_dihedrals & gather_impropers in python-scatter-gather.py

evoyiatzis committed:
fixing unit tests in python-scatter-gather.py

evoyiatzis commented:
@akohlmey I think I have addressed all points apart from (i) the unit tests for the C library calls and (ii) the unit tests for the new Fortran module calls. Could someone take over from here?

akohlmey commented:
> @akohlmey I think I have addressed all points apart from (i) the unit tests for the C library calls and (ii) the unit tests for the new Fortran module calls.

Thanks a lot. This is a big help.

> Could someone take over from here?

Sure thing. I'll take a look later today.

akohlmey committed:
Merge branch 'develop'

akohlmey committed:
whitespace

akohlmey committed:
add unit tests for c-library interface

akohlmey committed:
preparations for unit tests of the Fortran module

akohlmey committed:
Merge branch 'develop'

akohlmey committed:
correct syntax error

akohlmey committed:
Merge branch 'develop'

akohlmey committed:
convert Fortran version of lammps_gather_bonds() test to be similar to C version

akohlmey committed:
add tests for gathering angles, dihedrals, impropers from fortran

akohlmey committed:
use same main() function as with c-library interface.

akohlmey commented:
@evoyiatzis TODO list is now complete with adding the last few tests for the fortran interface.

akohlmey committed:
fix broken link

evoyiatzis commented:
@akohlmey great! thanks for the efforts to finalize this PR
